### PR TITLE
Iter25

### PR DESCRIPTION
**Pull request by Arcadian-Sky**

### CODE REVIEW

File: .github/codecov.yaml [hunk 1]
```diff
@@ -3,4 +3,9 @@ coverage:
   range: 55..100
   precision: 2
   status:
-    patch: no
\ No newline at end of file
+    patch: no
+  ignore:
+  - 'internal/server/mock/*'
```
Comment (arttet) [inline .github/codecov.yaml:8]: Please increase percent of code coverage up to 50%

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[submodule "proto/googleapi"]
+	path = proto/googleapi
```
Comment (arttet) [inline .gitmodules:2]: Please use `buf` https://buf.build/docs/introduction

File: cmd/server/main.go [hunk 3]
```diff
@@ -6,20 +6,27 @@ import (
...
 	"os/signal"
 	"syscall"
 	"time"
 
 	"net/http"
 
+	runtime "github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
 	_ "github.com/jackc/pgx/v5/stdlib"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
```
Comment (arttet) [inline cmd/server/main.go:20]: Please move it to another package. It should not be in `main` package.

File: proto/agent.proto [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+message MetricResponse {
+    string status = 1;
+}
+message MetricJSONRequest {
+    string json_string = 1;
+}
+
+service AgentService {
```
Comment (arttet) [inline proto/agent.proto:19]:
```suggestion
service AgentServiceV1 {
```

File: proto/agent.proto [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    rpc SendMetric (MetricRequest) returns (MetricResponse);
```
Comment (arttet) [inline proto/agent.proto:20]:
```suggestion
    rpc SendMetricV1(MetricRequestV1) returns (MetricResponseV1);
```

File: proto/agent.proto [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    string type = 1;
+    string name = 2;
+    string value = 3;
+}
+
+message MetricResponse {
+    string status = 1;
+}
+message MetricJSONRequest {
+    string json_string = 1;
```
Comment (arttet) [inline proto/agent.proto:16]: It should not be string. It doesn't allow you to pack data efficiently.

File: proto/agent.proto [hunk 7]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package agent;
+option go_package = "/internal/agent/generated/protoagent";
+
+message MetricRequest {
+    string type = 1;
```
Comment (arttet) [inline proto/agent.proto:7]: Please add validation.
https://github.com/bufbuild/protovalidate-go

File: proto/agent.proto [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    string name = 2;
+    string value = 3;
+}
+
+message MetricResponse {
+    string status = 1;
```
Comment (arttet) [inline proto/agent.proto:13]: What does it mean?

File: proto/agent.proto [hunk 9]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package agent;
+option go_package = "/internal/agent/generated/protoagent";
```
Comment (arttet) [inline proto/agent.proto:4]: It should not be in `internal`

File: proto/agent.proto (hunk 6 above)
Comment (arttet) [inline proto/agent.proto:16]: Please add `buf lint` in CI
https://github.com/bufbuild/buf-lint-action

File: proto/metrics.proto [hunk 10]
```diff
@@ -0,0 +1,129 @@
+//protoc -I $GOPATH/pkg/mod/github.com/googleapis/googleapis@latest/ -I . --grpc-gateway_out . --grpc-gateway_opt logtostderr=true --grpc-gateway_opt paths=source_relative service.proto
+//protoc -I proto/googleapi/ -I proto/. --grpc-gateway_out=. --grpc-gateway_opt=logtostderr=true --grpc-gateway_opt=paths=source_relative metrics.proto
+syntax = "proto3";
+
+package metrics;
```
Comment (arttet) [inline proto/metrics.proto:5]: Agent and Server should use the same proto files.

File: proto/metrics.proto [hunk 11]
```diff
@@ -0,0 +1,129 @@
+//protoc -I $GOPATH/pkg/mod/github.com/googleapis/googleapis@latest/ -I . --grpc-gateway_out . --grpc-gateway_opt logtostderr=true --grpc-gateway_opt paths=source_relative service.proto
```
Comment (arttet) [inline proto/metrics.proto:1]: Please use `buf`

File: proto/metrics.proto [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+// Ответ для обновления метрик через JSON
+message UpdateJSONMetricsResponse {
+    repeated Metric metrics = 1;
+}
+
+// Запрос для обновления метрики
+message UpdateMetricRequest {
+    string type = 1;
+    string name = 2;
+    string value = 3;
```
Comment (arttet) [inline proto/metrics.proto:31]: It should not be `string`

File: proto/metrics.proto [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    string value = 1;
+}
+
+// Ответ для получения метрики
+message GetJSONMetricsResponse {
+    repeated Metric metrics = 1;
+}
+
+// Запрос для проверки соединения с базой данных
+message PingDBRequest {}
```
Comment (arttet) [inline proto/metrics.proto:56]: Please use `google.protobuf.Empty`

File: proto/metrics.proto [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+message Metric {
+    string id = 1;
+    string type = 2;
+    int64 delta = 3;
+    double value = 4;
+}
+
+service MetricsService {
+    // Получает метрику
+    rpc GetMetric (Metric) returns (GetMetricResponse) {
```
Comment (arttet) [inline proto/metrics.proto:73]:
```suggestion
    rpc GetMetricV1 (GetMetricRequestV1) returns (GetMetricResponseV1) {
```

Comment (arttet) [reply proto/metrics.proto:73]: Please fix all methods of API

File: proto/metrics.proto [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+// Ответ для проверки соединения с базой данных
+message PingDBResponse {
+    string message = 1;
+}
+
+// Определение метрики
+message Metric {
+    string id = 1;
+    string type = 2;
+    int64 delta = 3;
```
Comment (arttet) [inline proto/metrics.proto:67]: Please use `oneof`

File: proto/metrics.proto [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Ответ для обновления метрики
+message UpdateMetricResponse {
+     repeated Metric metrics = 1;
+}
+
+// Запрос для получения метрики
+message GetMetricRequest {
+    string type = 1;
```
Comment (arttet) [inline proto/metrics.proto:41]: Please use `enum`

**Review by arttet — CHANGES_REQUESTED** [review]

File: proto/agent.proto (hunk 6 above)
Comment (Arcadian-Sky) [reply proto/agent.proto]: buf-lint-action deprecated added buf-action

**Review by arttet — CHANGES_REQUESTED** [review]
Please fix autotests
image

**Review by arttet — APPROVED** [review]

**Pull request merged**